File: packages/node/tests/parity.test.ts
// Sprint B D2 — cross-runtime parity test (Node leg).
//
// Asserts the three behavioral dimensions defined in
// docs/sdk/payload-format.md: header parity, payload structure
// parity, misconfiguration degradation parity. The other three
// runtimes have equivalent parity tests
// (ruby/spec/nurse_andrea/parity_spec.rb,
// python/tests/test_parity.py, go/nurseandrea/parity_test.go) that
// assert the same shape. The .github/workflows/sdk-parity.yml
// matrix runs all four; the suite is only meaningful if every leg
// passes.

import { configure, _resetForTests } from "../src/configuration"
import { client } from "../src/client"
import { deploy } from "../src/deploy"
import { SDK_LANGUAGE, SDK_VERSION } from "../src/version"

interface CapturedRequest {
  url: string
  headers: Record<string, string>
  body: any
}

const validConfig = () => ({
  orgToken: "org_parity_test_aaaaaaaaaaaaaaaaaaaa",
  workspaceSlug: "parity-test",
  environment: "development" as const,
  host: "http://parity.test",
  enabled: true,
  flushIntervalMs: 99999,
})

function installFetchCapture(): { captured: CapturedRequest[]; restore: () => void } {
  const captured: CapturedRequest[] = []
  const fetchMock = jest.fn(async (url: string, init: any) => {
    const headers: Record<string, string> = {}
    if (init?.headers) {
      for (const [k, v] of Object.entries(init.headers)) headers[k] = String(v)
    }
    const body = init?.body ? JSON.parse(init.body as string) : null
    captured.push({ url, headers, body })
    return { ok: true, status: 200, clone: () => ({ json: async () => ({}) }) }
  })
  const original = global.fetch
  global.fetch = fetchMock as any
  return { captured, restore: () => { global.fetch = original } }
}

describe("NurseAndrea SDK parity (Node)", () => {
  let capture: ReturnType<typeof installFetchCapture>

  beforeEach(() => {
    _resetForTests()
    configure(validConfig())
    client.resetRejectionState()
    jest.clearAllMocks()
    capture = installFetchCapture()
  })

  afterEach(() => {
    client.stop()
    capture.restore()
  })

  describe("Header parity", () => {
    it("emits the 5 canonical headers on /api/v1/ingest", async () => {
      client.enqueueLog({ level: "info", message: "x" })
      await (client as any).flush()
      const req = capture.captured.find(r => r.url.endsWith("/api/v1/ingest"))
      expect(req).toBeDefined()
      expect(req!.headers["Content-Type"]).toBe("application/json")
      expect(req!.headers["Authorization"]).toBe("Bearer org_parity_test_aaaaaaaaaaaaaaaaaaaa")
      expect(req!.headers["X-NurseAndrea-Workspace"]).toBe("parity-test")
      expect(req!.headers["X-NurseAndrea-Environment"]).toBe("development")
      expect(req!.headers["X-NurseAndrea-SDK"]).toBe(`${SDK_LANGUAGE}/${SDK_VERSION}`)
    })

    it("emits the 5 canonical headers on /api/v1/metrics", async () => {
      client.enqueueMetric({ name: "process.memory.rss", value: 1, unit: "bytes" })
      await (client as any).flush()
      const req = capture.captured.find(r => r.url.endsWith("/api/v1/metrics"))
      expect(req).toBeDefined()
      expect(req!.headers["Authorization"]).toBe("Bearer org_parity_test_aaaaaaaaaaaaaaaaaaaa")
      expect(req!.headers["X-NurseAndrea-Workspace"]).toBe("parity-test")
      expect(req!.headers["X-NurseAndrea-Environment"]).toBe("development")
      expect(req!.headers["X-NurseAndrea-SDK"]).toBe(`${SDK_LANGUAGE}/${SDK_VERSION}`)
    })

    it("emits the 5 canonical headers on /api/v1/deploy", async () => {
      await deploy({ version: "1.0.0" })
      const req = capture.captured.find(r => r.url.endsWith("/api/v1/deploy"))
      expect(req).toBeDefined()
      expect(req!.headers["Content-Type"]).toBe("application/json")
      expect(req!.headers["Authorization"]).toBe("Bearer org_parity_test_aaaaaaaaaaaaaaaaaaaa")
      expect(req!.headers["X-NurseAndrea-Workspace"]).toBe("parity-test")
      expect(req!.headers["X-NurseAndrea-Environment"]).toBe("development")
      expect(req!.headers["X-NurseAndrea-SDK"]).toBe(`${SDK_LANGUAGE}/${SDK_VERSION}`)
    })
  })

  describe("Payload structure parity", () => {
    it("log payload has canonical top-level + entry field names", async () => {
      client.enqueueLog({ level: "info", message: "parity", metadata: { k: "v" } })
      await (client as any).flush()
      const req = capture.captured.find(r => r.url.endsWith("/api/v1/ingest"))!
      expect(req.body).toEqual(
        expect.objectContaining({
          services: expect.any(Array),
          sdk_version: expect.any(String),
          sdk_language: "node",
          logs: expect.any(Array),
        })
      )
      const entry = req.body.logs[0]
      expect(entry).toEqual(
        expect.objectContaining({
          level: "info",
          message: "parity",
          occurred_at: expect.any(String),
          source: expect.any(String),
          payload: expect.any(Object),
        })
      )
    })

    it("metric payload has canonical top-level + entry field names", async () => {
      client.enqueueMetric({ name: "process.memory.rss", value: 1, unit: "bytes" })
      await (client as any).flush()
      const req = capture.captured.find(r => r.url.endsWith("/api/v1/metrics"))!
      expect(req.body).toEqual(
        expect.objectContaining({
          sdk_version: expect.any(String),
          sdk_language: "node",
          metrics: expect.any(Array),
        })
      )
      const entry = req.body.metrics[0]
      expect(entry).toEqual(
        expect.objectContaining({
          name: "process.memory.rss",
          value: 1,
          unit: "bytes",
          occurred_at: expect.any(String),
          tags: expect.any(Object),
        })
      )
    })
  })

  describe("Misconfig degradation parity", () => {
    it("missing orgToken does not throw and skips enqueue", async () => {
      _resetForTests()
      // Configure without orgToken — direct internal write since
      // configure() validates required fields. The
      // is-enabled check inside enqueueLog short-circuits.
      configure({
        orgToken: "",
        workspaceSlug: "parity-test",
        environment: "development",
        host: "http://parity.test",
        enabled: true,
      } as any)
      expect(() => client.enqueueLog({ level: "info", message: "x" })).not.toThrow()
      // No fetch call attempted because isEnabled() short-circuits
      // when the SDK is misconfigured at the client layer.
      const calls = capture.captured.length
      await (client as any).flush()
      expect(capture.captured.length).toBe(calls)
    })
  })
})

File: packages/go/parity_test.go
package nurseandrea_test

// Sprint B D2 — cross-runtime parity test (Go leg).
//
// Asserts the three behavioral dimensions defined in
// docs/sdk/payload-format.md: header parity, payload structure
// parity, misconfiguration degradation parity. The other three
// runtimes have equivalent parity tests
// (ruby/spec/nurse_andrea/parity_spec.rb,
// node/tests/parity.test.ts, python/tests/test_parity.py) that
// assert the same shape. The .github/workflows/sdk-parity.yml
// matrix runs all four; the suite is only meaningful if every leg
// passes.

import (
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"regexp"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/narteyb/nurse-andrea-sdks/packages/go/nurseandrea"
)

type captured struct {
	mu       sync.Mutex
	requests []*http.Request
	bodies   [][]byte
}

func (c *captured) record(r *http.Request) {
	c.mu.Lock()
	defer c.mu.Unlock()
	body, _ := io.ReadAll(r.Body)
	c.requests = append(c.requests, r)
	c.bodies = append(c.bodies, body)
}

func (c *captured) waitFor(t *testing.T, suffix string, timeout time.Duration) (*http.Request, []byte) {
	t.Helper()
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		c.mu.Lock()
		for i, r := range c.requests {
			if strings.HasSuffix(r.URL.Path, suffix) {
				body := c.bodies[i]
				c.mu.Unlock()
				return r, body
			}
		}
		c.mu.Unlock()
		time.Sleep(10 * time.Millisecond)
	}
	t.Fatalf("no POST to %s within %s", suffix, timeout)
	return nil, nil
}

func parityHost(t *testing.T, c *captured) *httptest.Server {
	t.Helper()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		c.record(r)
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("{}"))
	}))
	t.Cleanup(srv.Close)
	return srv
}

func parityConfigure(t *testing.T, host string) {
	t.Helper()
	resetConfig()
	err := nurseandrea.Configure(nurseandrea.Config{
		OrgToken:      "org_parity_test_aaaaaaaaaaaaaaaaaaaa",
		WorkspaceSlug: "parity-test",
		Environment:   "development",
		Host:          host,
		Enabled:       nurseandrea.BoolPtr(true),
		BatchSize:     1,
	})
	if err != nil {
		t.Fatalf("Configure: %v", err)
	}
}

// ─── Header parity ─────────────────────────────────────────────

func TestParityHeadersIngest(t *testing.T) {
	c := &captured{}
	srv := parityHost(t, c)
	parityConfigure(t, srv.URL)

	nurseandrea.GetClient().EnqueueLog("info", "x", nil)
	req, _ := c.waitFor(t, "/api/v1/ingest", 2*time.Second)
	checkCanonicalHeaders(t, req, "go")
}

func TestParityHeadersMetrics(t *testing.T) {
	c := &captured{}
	srv := parityHost(t, c)
	parityConfigure(t, srv.URL)

	nurseandrea.GetClient().EnqueueMetric("process.memory.rss", 1.0, "bytes", nil)
	req, _ := c.waitFor(t, "/api/v1/metrics", 2*time.Second)
	checkCanonicalHeaders(t, req, "go")
}

func TestParityHeadersDeploy(t *testing.T) {
	c := &captured{}
	srv := parityHost(t, c)
	parityConfigure(t, srv.URL)

	nurseandrea.Deploy("1.0.0")
	req, _ := c.waitFor(t, "/api/v1/deploy", 2*time.Second)
	// Sprint B D2 added X-NurseAndrea-SDK to Go's deploy headers.
	checkCanonicalHeaders(t, req, "go")
}

func checkCanonicalHeaders(t *testing.T, r *http.Request, lang string) {
	t.Helper()
	if got := r.Header.Get("Content-Type"); got != "application/json" {
		t.Errorf("Content-Type: got %q want application/json", got)
	}
	if got := r.Header.Get("Authorization"); got != "Bearer org_parity_test_aaaaaaaaaaaaaaaaaaaa" {
		t.Errorf("Authorization: got %q", got)
	}
	if got := r.Header.Get("X-NurseAndrea-Workspace"); got != "parity-test" {
		t.Errorf("X-NurseAndrea-Workspace: got %q", got)
	}
	if got := r.Header.Get("X-NurseAndrea-Environment"); got != "development" {
		t.Errorf("X-NurseAndrea-Environment: got %q", got)
	}
	sdk := r.Header.Get("X-NurseAndrea-SDK")
	matched, _ := regexp.MatchString(`^`+lang+`/\d+\.\d+\.\d+$`, sdk)
	if !matched {
		t.Errorf("X-NurseAndrea-SDK: got %q, expected %s/<semver>", sdk, lang)
	}
}

// ─── Payload structure parity ─────────────────────────────────

func TestParityLogPayloadStructure(t *testing.T) {
	c := &captured{}
	srv := parityHost(t, c)
	parityConfigure(t, srv.URL)

	nurseandrea.GetClient().EnqueueLog("info", "parity", map[string]interface{}{"k": "v"})
	_, body := c.waitFor(t, "/api/v1/ingest", 2*time.Second)

	var parsed map[string]interface{}
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("body not JSON: %v", err)
	}
	for _, key := range []string{"services", "sdk_version", "sdk_language", "logs"} {
		if _, ok := parsed[key]; !ok {
			t.Errorf("top-level missing %q", key)
		}
	}
	if parsed["sdk_language"] != "go" {
		t.Errorf("sdk_language: got %v want go", parsed["sdk_language"])
	}
	logs := parsed["logs"].([]interface{})
	entry := logs[0].(map[string]interface{})
	for _, key := range []string{"level", "message", "occurred_at", "source", "payload"} {
		if _, ok := entry[key]; !ok {
			t.Errorf("log entry missing %q (got keys: %v)", key, keys(entry))
		}
	}
}

func TestParityMetricPayloadStructure(t *testing.T) {
	c := &captured{}
	srv := parityHost(t, c)
	parityConfigure(t, srv.URL)

	nurseandrea.GetClient().EnqueueMetric("process.memory.rss", 1.0, "bytes", nil)
	_, body := c.waitFor(t, "/api/v1/metrics", 2*time.Second)

	var parsed map[string]interface{}
	if err := json.Unmarshal(body, &parsed); err != nil {
		t.Fatalf("body not JSON: %v", err)
	}
	for _, key := range []string{"sdk_version", "sdk_language", "metrics"} {
		if _, ok := parsed[key]; !ok {
			t.Errorf("top-level missing %q", key)
		}
	}
	if parsed["sdk_language"] != "go" {
		t.Errorf("sdk_language: got %v want go", parsed["sdk_language"])
	}
	metrics := parsed["metrics"].([]interface{})
	entry := metrics[0].(map[string]interface{})
	for _, key := range []string{"name", "value", "unit", "occurred_at", "tags"} {
		if _, ok := entry[key]; !ok {
			t.Errorf("metric entry missing %q (got keys: %v)", key, keys(entry))
		}
	}
}

func keys(m map[string]interface{}) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	return out
}

// ─── Misconfig degradation parity ─────────────────────────────

func TestParityMisconfigReturnsError(t *testing.T) {
	// Go's idiom is "return error" rather than "raise" — Configure
	// signals misconfiguration via err != nil. The cross-runtime
	// parity contract (docs/sdk/payload-format.md §6) calls for "no
	// raise at SDK boot" — returning an error qualifies as not
	// raising. Callers are expected to inspect the error and either
	// log-and-continue or fail-fast at their discretion.
	resetConfig()
	err := nurseandrea.Configure(nurseandrea.Config{
		// No OrgToken.
		WorkspaceSlug: "parity-test",
		Environment:   "development",
	})
	if err == nil {
		t.Fatal("expected error for missing OrgToken")
	}
	if !strings.Contains(err.Error(), "OrgToken") {
		t.Errorf("error message does not name OrgToken: %v", err)
	}
}
